hail/src/main/scala/is/hail/expr/ir/functions/IntervalFunctions.scala

```diff
@@ -137,41 +137,114 @@ object IntervalFunctions extends RegistryFunctions {
+        val point = _point.asBaseStruct.memoize(cb, "point")
+
+        val left = leftPartitionEndpoint.loadField(cb, 0).get(cb).asBaseStruct.memoize(cb, "partitionInterval_left")
+        val leftLen = cb.newLocal[Int]("partitionInterval_leftlen", leftPartitionEndpoint.loadField(cb, 1).get(cb).asInt.intCode(cb))
+
+        val c = cb.newLocal[Int]("partitionInterval_c", 0)
+        (0 until left.st.size).foreach { idx =>
+          cb.ifx(c.ceq(0) && const(idx) < leftLen, {
+            val leftField = EmitCode.fromI(cb.emb)(cb => left.loadField(cb, idx))
+            val pointField = EmitCode.fromI(cb.emb)(cb => point.loadField(cb, idx))
+            cb.assign(c, cb.emb.ecb.getOrderingFunction(leftField.st, pointField.st, CodeOrdering.Compare())
+              .apply(cb, leftField, pointField))
+          })
         }
 
-        bindIRs(point, GetField(interval, "left"), GetField(interval, "right")) { case Seq(point, l, r) =>
+        val isContained = cb.newLocal[Boolean]("partitionInterval_b")
+        cb.ifx(c.ceq(0),
+          cb.assign(isContained, _containsStart.asBoolean.boolCode(cb)),
+          cb.assign(isContained, c < 0))
 
+        primitive(isContained)
+    }
 
-          val gtEqLeft = bindIR(compareStructs(l, point)) { lc =>
-            (lc <= 0) && ((lc < 0) || GetField(interval, "includesLeft"))
-          }
+    registerSCode3("partitionIntervalEndpointLessThan", endpointT, tv("T"), TBoolean, TBoolean, (_, _, _, _) => SBoolean) {
+      case (_, cb, _, _rightEndpoint, _point, _containsEnd, _) =>
 
-          val ltEqRight = bindIR(compareStructs(r, point)) { rc =>
-            (rc >= 0) && ((rc > 0) || GetField(interval, "includesRight"))
-          }
-          gtEqLeft && ltEqRight
+        val rightPartitionEndpoint = _rightEndpoint.asBaseStruct.memoize(cb, "partitionEndpoint")
+        val point = _point.asBaseStruct.memoize(cb, "point")
+
+        val right = rightPartitionEndpoint.loadField(cb, 0).get(cb).asBaseStruct.memoize(cb, "partitionInterval_right")
+        val rightLen = cb.newLocal[Int]("partitionInterval_rightlen", rightPartitionEndpoint.loadField(cb, 1).get(cb).asInt.intCode(cb))
+
+        val c = cb.newLocal[Int]("partitionInterval_c", 0)
+        (0 until right.st.size).foreach { idx =>
+          cb.ifx(c.ceq(0) && const(idx) < rightLen, {
+            val rightField = EmitCode.fromI(cb.emb)(cb => right.loadField(cb, idx))
+            val pointField = EmitCode.fromI(cb.emb)(cb => point.loadField(cb, idx))
+            cb.assign(c, cb.emb.ecb.getOrderingFunction(rightField.st, pointField.st, CodeOrdering.Compare())
+              .apply(cb, rightField, pointField))
+          })
         }
+
+        val isContained = cb.newLocal[Boolean]("partitionInterval_b")
+        cb.ifx(c.ceq(0),
+          cb.assign(isContained, _containsEnd.asBoolean.boolCode(cb)),
+          cb.assign(isContained, c > 0))
+
+        primitive(isContained)
+    }
+
+
+    registerSCode2("partitionIntervalContains",
```

Comment:
This has the same semantics as the previous IR function, but I rewrote it as an SCode function for clarity and method size considerations.